ring.go

```diff
@@ -220,25 +222,94 @@ type ringShards struct {
 }
 
 func newRingShards(opt *RingOptions) *ringShards {
-	shards := make(map[string]*ringShard, len(opt.Addrs))
+	shards, list := createRingShards(opt, opt.Addrs)
+
+	c := &ringShards{
+		opt: opt,
+
+		shards: shards,
+		list:   list,
+	}
+	c.rebalance()
+
+	return c
+}
+
+func createRingShards(opt *RingOptions, addrs map[string]string) (map[string]*ringShard, []*ringShard) {
+	shards := make(map[string]*ringShard, len(addrs))
 	list := make([]*ringShard, 0, len(shards))
 
-	for name, addr := range opt.Addrs {
+	for name, addr := range addrs {
 		shard := newRingShard(opt, name, addr)
 		shards[name] = shard
 
 		list = append(list, shard)
 	}
+	return shards, list
+}
 
-	c := &ringShards{
-		opt: opt,
+func (c *ringShards) SetAddrs(addrs map[string]string) {
+	var currentRingshards map[string]*ringShard
+	toDelete := make(map[string]*ringShard)
+	newShards := make(map[string]*ringShard)
 
-		shards: shards,
-		list:   list,
+	c.mu.RLock()
+	if !c.closed {
+		currentRingshards = c.shards
+	}
+	c.mu.RUnlock()
+
+	// delete already connected shard addresses
+	for k, currentRingshard := range currentRingshards {
+		if v, ok := addrs[k]; ok {
+
+			if currentRingshard.addr == v {
+				// same - reuse old addr
+				delete(addrs, k)
+			} else {
+				// replace addr
+				toDelete[k] = currentRingshard
+			}
+		} else {
+			// decrease: stop using these
+			toDelete[k] = currentRingshard
+		}
+	}
+
+	// create new connections
+	for name, addr := range addrs {
+		newShards[name] = newRingShard(c.opt, name, addr)
+	}
+	// likely needs to be in critical section, if not we can use the loop before and drop "newShards"
+	for name, shard := range newShards {
+		currentRingshards[name] = shard
+	}
```

Comment:
This two loops could be combined to eliminate `newShards` altogether.

Reply:
yes, that's on my list, but maybe we need to guard the assignment and then it might make sense to have both separated to not block the map while creating connection pools.